launch_ros/launch_ros/actions/node.py

```diff
@@ -0,0 +1,187 @@
+        :param: node_executable the name of the executable to find
+        :param: node_name the name of the node
+        :param: node_namespace the ros namespace for this Node
+        :param: remappings map of 'to' and 'from' strings to be passed to the
+            node as ROS remapping rules
+        :param: arguments list of extra arguments for the node
+        """
+        cmd = [ExecutableInPackage(package=package, executable=node_executable)]
+        cmd += [] if arguments is None else arguments
+        # Reserve space for ros specific arguments.
+        # The substitutions will get expanded when the action is executed.
+        ros_args_index = 0
+        if node_name is not None:
+            cmd += [LocalSubstitution('ros_specific_arguments[{}]'.format(ros_args_index))]
+            ros_args_index += 1
+        if node_namespace is not None:
+            cmd += [LocalSubstitution('ros_specific_arguments[{}]'.format(ros_args_index))]
+            ros_args_index += 1
+        if remappings is not None:
+            for k in remappings.keys():
+                cmd += [LocalSubstitution('ros_specific_arguments[{}]'.format(ros_args_index))]
+                ros_args_index += 1
+        super().__init__(cmd=cmd, **kwargs)
+        self.__package = package
+        self.__node_executable = node_executable
+        self.__node_name = node_name
+        self.__node_namespace = node_namespace
+        self.__remappings = {} if remappings is None else remappings
+        self.__arguments = arguments
+
+        self.__expanded_node_name = '<node_name_unspecified>'
+        self.__expanded_node_namespace = '/'
+        self.__final_node_name: Optional[Text] = None
+        self.__expanded_remappings: Optional[Dict[Text, Text]] = None
+
+        self.__substitutions_performed = False
+
+    @property
+    def node_name(self):
+        """Getter for node_name."""
+        if self.__final_node_name is None:
+            raise RuntimeError("cannot access 'node_name' before executing action")
+        return self.__final_node_name
+
+    def _perform_substitutions(self, context: LaunchContext) -> None:
+        try:
+            if self.__substitutions_performed:
+                return
+            self.__substitutions_performed = True
+            if self.__node_name is not None:
+                self.__expanded_node_name = perform_substitutions(
+                    context, normalize_to_list_of_substitutions(self.__node_name))
+                validate_node_name(self.__node_name)
+            self.__expanded_node_name.lstrip('/')
+            if self.__node_namespace is not None:
+                self.__expanded_node_namespace = perform_substitutions(
+                    context, normalize_to_list_of_substitutions(self.__node_namespace))
+            if not self.__expanded_node_namespace.startswith('/'):
+                self.__expanded_node_namespace = '/' + self.__expanded_node_namespace
```

Comment:
what is the point of adding a leading forward slash if the next line does `lstrip('/')`?

Reply:
No idea, 03e9d07